### ISSUE TITLE
Compilation error with -DTORRENT_NO_DEPRECATE

### ISSUE THREAD

**Issue by evsh** [arvidn/libtorrent#3009]
**Please provide the following information**

libtorrent version (or branch): master

platform/architecture: Gentoo Linux/amd64

compiler and compiler version: GCC 8

Here is the error message:
```
/usr/include/libtorrent/create_torrent.hpp:415:19: error: no matching function for call to ‘set_piece_hashes(libtorrent::create_torrent&, const string&, BitTorrent::TorrentCreatorThread::run()::<lambda(int)>&, libtorrent::error_code&)’
   set_piece_hashes(t, p, f, ec);
   ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
/usr/include/libtorrent/create_torrent.hpp:398:22: note: candidate: ‘void libtorrent::set_piece_hashes(libtorrent::create_torrent&, const string&, const std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag>)>&, libtorrent::error_code&)’
  TORRENT_EXPORT void set_piece_hashes(create_torrent& t, std::string const& p
                      ^~~~~~~~~~~~~~~~
/usr/include/libtorrent/create_torrent.hpp:398:22: note:   no known conversion for argument 3 from ‘BitTorrent::TorrentCreatorThread::run()::<lambda(int)>’ to ‘const std::function<void(libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag>)>&’
/usr/include/libtorrent/create_torrent.hpp:400:14: note: candidate: ‘void libtorrent::set_piece_hashes(libtorrent::create_torrent&, const string&, libtorrent::error_code&)’
  inline void set_piece_hashes(create_torrent& t, std::string const& p, error_code& ec)
              ^~~~~~~~~~~~~~~~
/usr/include/libtorrent/create_torrent.hpp:400:14: note:   candidate expects 3 arguments, 4 provided
/usr/include/libtorrent/create_torrent.hpp:405:14: note: candidate: ‘void libtorrent::set_piece_hashes(libtorrent::create_torrent&, const string&)’
  inline void set_piece_hashes(create_torrent& t, std::string const& p)
              ^~~~~~~~~~~~~~~~
/usr/include/libtorrent/create_torrent.hpp:405:14: note:   candidate expects 2 arguments, 4 provided
/usr/include/libtorrent/create_torrent.hpp:412:7: note: candidate: ‘template<class Fun> void libtorrent::set_piece_hashes(libtorrent::create_torrent&, const string&, Fun)’
  void set_piece_hashes(create_torrent& t, std::string const& p, Fun f)
       ^~~~~~~~~~~~~~~~
/usr/include/libtorrent/create_torrent.hpp:412:7: note:   template argument deduction/substitution failed:
/usr/include/libtorrent/create_torrent.hpp:415:19: note:   candidate expects 3 arguments, 4 provided
   set_piece_hashes(t, p, f, ec);
   ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
```
The code fragment from the `create_torrent.hpp` file:
```c++
template <class Fun>
void set_piece_hashes(create_torrent& t, std::string const& p, Fun f)
{
	error_code ec;
	set_piece_hashes(t, p, f, ec); // error at this line
	if (ec) throw system_error(ec);
}
```

**Comment by arvidn** [comment]
ah, yes. This signature has changed. It no longer passes in an `int`, it passes in a `piece_index_t` to the callback function

**Comment by arvidn** [comment]
I mean, the previous signature has been deprecated. Perhaps this is a poor way of exposing it. Maybe there should really be a separate `set_piece_hashes()` function for the new callback signature.

**Comment by evsh** [comment]
I don't know your intentions (hence the report), but it is certainly surprising to find the function signature requirement inside of the libtorrent function, but not in its declaration.  An `std::function` with the required signature at the very first libtorrent API call would be desirable.

**Comment by arvidn** [comment]
well, it's a template and we don't have concepts yet...

**Comment by evsh** [comment]
But it forwards the call to a function that accepts `std::function` only, right? What's the point in the template parameter than?

**Comment by arvidn** [comment]
that's a good question. I suppose it may have been refactored into this shape at some point

**Comment by arvidn** [comment]
does this look ok? https://github.com/arvidn/libtorrent/pull/3013
it won't fix the build error, but probably make for a better error.
You do get deprecation warnings when building without the `TORRENT_NO_DEPRECATE` define, right?

**Comment by evsh** [comment]
Yes, thanks, that puts it straight.

**Comment by evsh** [comment]
Closed via #3013.
